### PR TITLE
feat(0001): Variables

### PR DESCRIPTION
**Pull request by lino-levan**

### CODE REVIEW

**Comment by nmarks413** [comment]
I think many people are more receptive to const rather than let mut. I quite like the rust way, but if you are targeting this at JS devs, then familiarity would be quite useful. Zig provides a good example of how this would work, and as a lot of the philosophies seem lifted from it, it might be useful to carry that along.

**Comment by rozjin** [comment]
the language is going to be a rusty js with extensive differences from JavaScript and it's primary goal is to be Typescript 2 without the incrimental change so I don't see the point in const. Additionally let vs let mut is very obvious in code and using modifiers for keywords is a nice paradigm I'd like to keep for the future.

**Comment by lino-levan** [comment]
> I think many people are more receptive to const rather than let mut. I quite like the rust way, but if you are targeting this at JS devs, then familiarity would be quite useful. Zig provides a good example of how this would work, and as a lot of the philosophies seem lifted from it, it might be useful to carry that along.

I've discussed with a lot of people internally the answer is basically: `let mut` looks to be the best option. At first it seems arbitrary. Why would I want two keywords instead of one keyword to do the same thing? It turns out that this is critical to future language components. Let's say we went with `var` and `const` like zig. How would functions with immutable / mutable arguments work?
```rust
fn <name>(var x: <type>, const y: <type>) {

}
```
looks and feels bad, but unless we want to define more keywords, this is kind of the only way to do it. Let's suppose though that you keep insisting. We could do
```rust
fn <name>(var x: <type>, y: <type>) {

}
```
but from a beginner perspective, what the heck is `var` doing there. What does `var` represent? I would struggle to read that. `mut` is a much clearer keyword imo.

But even then, let's suppose that you think this is fine. New issue for you: object destructuring. What if you want some destructured values to be mutable and some not to be mutable. With the `mut` keyword, it's pretty clear:
```rs
let (x, mut y, z) = <value>;
```
It starts to be very dicey otherwise, and it becomes very unclear how to do this. You could do
```
const (x, var y, z) = <value>;
```
but that looks incredibly gross. You could also do
```
(const x, var y, const z) = <value>;
```
which looks potentially even grosser. The reason why zig didn't pick `let mut` is because they didn't have destructuring until just recently. Even then, it's just array destructuring. Even then, it looks really gross. We can do better.

**Comment by lino-levan** [comment]
This RFC has been limited in scope to only include immutable variables due to some conflicts. Mutable variables will be discussed later.

File: text/0001-variables.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+The variable name can be any valid unicode except for the following conditions:
+
+- Variables may not start with numerical digits
+- Variables may not have any whitespace characters
+- Variables may not contain with any of the following characters as they are
+  reserved for future language use: "'<>(){}[]\|/?,.=+-~`!@#$%^&*
+
+```blight
+// Valid
+let x: <type> = <value>;
```
Comment (OlexG) [inline text/0001-variables.md:55]: Is there a reason let is used and not const, as variables are by default immutable? Or is this part of the philosophy, so the immutability doesn't have to be implied?

**Review by OlexG — APPROVED** [review]
Me likey likey, simple and to the point

File: text/0001-variables.md (hunk 1 above)
Comment (lino-levan) [reply text/0001-variables.md:55]: I think it's part of the philosophy. This was... hotly debated. We're going with `let` for now and we'll revisit that decision later.

**Pull request merged**